### PR TITLE
add registrations_require_3pid and allow_local_3pids

### PR DESCRIPTION
**Pull request by ara4n**
Adds `allow_local_3pids`: to let homeservers specify a whitelist for the format of 3PIDs that users are allowed to register with or add to their HS accounts.

It does not directly impact the 3PIDs which a user can bind to their account on their IS.

It also adds `registrations_require_3pid`, to determine which registration flows are advertised by the HS in case an admin wants to force users to specify a given 3PID when they sign up.

Typically useful for stopping people from registering with non-work emails.

### CODE REVIEW

File: synapse/config/registration.py [hunk 1]
```diff
@@ -31,6 +31,7 @@ def read_config(self, config):
                 strtobool(str(config["disable_registration"]))
             )
 
+        self.registrations_require_3pid = config.get("registrations_require_3pid", [])
```
Comment (turt2live) [inline synapse/config/registration.py]: This doesn't seem to affect just registrations and could be misleading to some people.

**Comment by ara4n** [comment]
Rewritten based on the above to:

 * [ ] split config options into allowed_local_3pids and registrations_require_3pid
 * [ ] simplify and comment logic for picking registration flows
 * [ ] fix docstring and move check_3pid_allowed into a new util module
 * [ ] use check_3pid_allowed everywhere
 * [ ] add trailing commas when we wrap args...

File: synapse/util/threepids.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        hs (synapse.server.HomeServer): server
+        medium (str): 3pid medium - e.g. email, msisdn
+        address (str): address within that medium (e.g. "wotan@matrix.org")
+            msisdns need to first have been canonicalised
+    """
+
+    if hs.config.allowed_local_3pids:
+        for constraint in hs.config.allowed_local_3pids:
+            logger.debug("Checking 3PID %s (%s) against %s (%s)" % (
+                address, medium, constraint['pattern'], constraint['medium']
```
Comment (erikjohnston) [inline synapse/util/threepids.py]: Don't need explicit string formatting, logger.debug does it for you.

File: synapse/rest/client/v2_alpha/register.py [hunk 3]
```diff
@@ -305,31 +316,70 @@ def on_POST(self, request):
...
+
+        if (
+            auth_result and
+            (
+                LoginType.EMAIL_IDENTITY in auth_result or
+                LoginType.EMAIL_MSISDN in auth_result
+            )
+        ):
+            medium = auth_result[LoginType.EMAIL_IDENTITY].threepid['medium']
+            address = auth_result[LoginType.EMAIL_IDENTITY].threepid['address']
```
Comment (erikjohnston) [inline synapse/rest/client/v2_alpha/register.py]: Doesn't this need to check both msisdn and email?

File: synapse/config/registration.py [hunk 4]
```diff
@@ -52,6 +54,23 @@ def default_config(self, **kwargs):
         # Enable registration for new users.
         enable_registration: False
 
+        # The user must provide all of the below types of 3PID when registering.
+        #
+        # registrations_require_3pid:
+        #     - email
+        #     - msisdn
```
Comment (erikjohnston) [inline synapse/config/registration.py]: Personally I would have the two config option entirely separate, as this doesn't support the use case of "require a work email but let people bind personal emails"

File: synapse/util/threepids.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+def check_3pid_allowed(hs, medium, address):
+    """Checks whether a given format of 3PID is allowed to be used on this HS
+
+    Args:
+        hs (synapse.server.HomeServer): server
+        medium (str): 3pid medium - e.g. email, msisdn
+        address (str): address within that medium (e.g. "wotan@matrix.org")
+            msisdns need to first have been canonicalised
+    """
```
Comment (erikjohnston) [inline synapse/util/threepids.py]: Should have a:

```
Returns:
    bool: ...
```

especially since I think there are other `check_` functions raise rather than return a bool

File: synapse/rest/client/v2_alpha/register.py (hunk 3 above)
Comment (ara4n) [reply synapse/rest/client/v2_alpha/register.py]: oops, thanks

File: synapse/config/registration.py (hunk 4 above)
Comment (ara4n) [reply synapse/config/registration.py]: we're not trying to solve for that use case here, plus having talked to @dbkr it's hard to do anyway - as /account/3pid goes and adds the email to your account (letting you log in with it) as well as optionally binding it on the identity server... so you can't separate the two easily anyway.

Hence me splitting it into:
 * What 3PID media do we mandate at registration?
 * What format 3PIDs are allowed to be added to accounts on this HS?
...with no restrictions on formats valid on ISes.

Comment (ara4n) [reply synapse/config/registration.py]: btw, this is a good example of a review comment where I'm staring at it thinking "does he really want me to rewrite it again? or is he just voicing a random comment?".  It might be worth clearly spelling out which comments actually block accepting the PR from your POV, otherwise the committer has no option than assume everything is a request for change...

Comment (erikjohnston) [reply synapse/config/registration.py]: Ok, in this case it was an observation that this feels odd and questioning whether this is the right thing to do. The synapse review process is not just about whether functional changes are right, but if we should even be doing it like that.

File: synapse/config/registration.py (hunk 1 above)
Comment (ara4n) [reply synapse/config/registration.py]: this convo ended up in IRL as dave presumably missed this comment: the impact is:
 * There are three request tokens atm for verifying 3PIDs:
   * registration
   * password reset
   * adding a 3pid to your account for use in login (which in turn tries to go bind it on the IS if asked)
  * this PR restricts the format of all three with allowed_local_3pids
  * it does not limit IS bindings (other than in a weak manner given in practice the account/3pid api is used by clients today to trigger bindings)

**Comment by ara4n** [comment]
> I'm confused what this is trying to do. Is this:
> 1. requiring that users register with a 3pid in a whitelist? or
> 2. stopping users from adding 3pids outside a whitelist?

Both. It's trying to limit both the 3PIDs which a user can use to register, and add to their HS account.  The main focus is on requiring the user has a valid email address for a domain at the point that they register; however if this restriction is in place, it seems unwise to let them then go and add other emails to let them log into the account (such that they might still be able to login and reset passwords etc, even if their official address is no more).

In other words: it feels needlessly confusing to expose a different whitelist for 3PID formats which may be added to an account, than that which is used to register in the first place.

**Comment by ara4n** [comment]
> I think this does what you want, but I think we need to be a lot more explicit that we're requiring and restricting 3pids here. I think it may be better to split up the config (and PR) into "require 3pid on register" and "disallow 3pids outside a whitelist"

Totally agreed, which is why I rewrote the PR to do this.

**Comment by ara4n** [comment]
@erikjohnston ptal again if you can face it.

**Comment by richvdh** [comment]
@erikjohnston did you mean to assign this to me?

**Comment by erikjohnston** [comment]
> @erikjohnston did you mean to assign this to me?

woops, sorry, must have been muscle memory

**Comment by ara4n** [comment]
thanks :)
